### PR TITLE
Changes for CD pipeline

### PR DESCRIPTION
**Pull request by anirudhacharya**
### Description of changes:
Changes for CD pipeline

#### Style and formatting:

I have run `pre-commit install` to ensure that auto-formatting happens with every commit.

#### Issue number, if available

By submitting this pull request, I confirm that you can use, modify, copy, and redistribute this contribution, under the terms of your choice.

### CODE REVIEW

File: config/buildspec_build_wheel.yml [hunk 1]
```diff
@@ -0,0 +1,38 @@
...
+        - sudo apt-get update -qq -o=Dpkg::Use-Pty=0
+        - sudo apt-get install unzip -qq -o=Dpkg::Use-Pty=0
+        - cd $CODEBUILD_SRC_DIR && chmod +x config/protoc_downloader.sh && ./config/protoc_downloader.sh
+        - pip install -U pip
+        - pip install -q pytest wheel pyYaml pytest-html pre-commit
+        - pip uninstall -y boto3 && pip uninstall -y aiobotocore && pip uninstall -y botocore
+
+  build:
+    commands:
+      - cd $CODEBUILD_SRC_DIR  && if [ -z "$CORE_COMMIT" ]; then export CORE_COMMIT=$(git log -1 --pretty=%h); fi
```
Comment (Vikas-kum) [inline config/buildspec_build_wheel.yml]: Let's put env variables in ENV section at top . Example
env:
  variables:
    ## comment, 
    CORE_COMMIT: latest
   RULES_COMMIT

File: config/buildspec_zero_code_change.yml [hunk 2]
```diff
@@ -18,9 +18,8 @@ env:
 phases:
   install:
     commands:
-        -  . config/change_branch.sh
```
Comment (Vikas-kum) [inline config/buildspec_zero_code_change.yml]: why was this removed?

We may want to start build selectively. Say, I want to run TF build on certain commit/branch, 
Is above possible if you remove this.

File: config/buildspec_zero_code_change.yml [hunk 3]
```diff
@@ -30,14 +29,10 @@ phases:
   build:
     commands:
       - cd $CODEBUILD_SRC_DIR && chmod +x config/install_smdebug.sh && chmod +x config/check_smdebug_install.sh && ./config/install_smdebug.sh && cd ..
-      - cd $RULES_CODEBUILD_SRC_DIR && python setup.py bdist_wheel --universal && pip install dist/*.whl && cd ..
-      - cd $CODEBUILD_SRC_DIR  && chmod +x config/tests.sh && PYTHONPATH=. ./config/tests.sh  && mkdir -p upload/$CURRENT_COMMIT_PATH/wheels && cp ./dist/*.whl upload/$CURRENT_COMMIT_PATH/wheels && cd ..
-      - cd $RULES_CODEBUILD_SRC_DIR && chmod +x config/tests.sh && mkdir -p upload/$CURRENT_COMMIT_PATH/wheels && PYTHONPATH=. ./config/tests.sh  && cp ./dist/*.whl upload/$CURRENT_COMMIT_PATH/wheels && cd ..
```
Comment (Vikas-kum) [inline config/buildspec_zero_code_change.yml]: why removed?

File: config/buildspec_zero_code_change.yml [hunk 4]
```diff
@@ -30,14 +29,10 @@ phases:
...
+      - cd $CODEBUILD_SRC_DIR  && chmod +x config/tests.sh && PYTHONPATH=. ./config/tests.sh
+      - cd $CODEBUILD_SRC_DIR_RULES && chmod +x config/tests.sh
 
   post_build:
     commands:
-      - . $CODEBUILD_SRC_DIR/config/upload_on_end.sh
-      - rm -rf $CODEBUILD_SRC_DIR/upload/$CURRENT_COMMIT_PATH
-      - rm -rf $RULES_CODEBUILD_SRC_DIR/upload/$CURRENT_COMMIT_PATH
```
Comment (Vikas-kum) [inline config/buildspec_zero_code_change.yml]: why removed, I guess upload on end also uploads logs

File: config/install_smdebug.sh [hunk 5]
```diff
@@ -3,11 +3,13 @@ set -ex
...
 
 
+export SMDEBUG_S3_BINARY="s3://smdebug-nightly-binaries/$(date +%F)/"
 cd $CODEBUILD_SRC_DIR
 # you can provide bip binary as s3 path in the build environment
 if [ "$SMDEBUG_S3_BINARY" ]; then
+  echo "Installing smdebug and smdebug_rules from pre-generated pip wheels at $SMDEBUG_S3_BINARY"
   mkdir -p s3_pip_binary
-  aws s3 cp "$SMDEBUG_S3_BINARY" s3_pip_binary
+  aws s3 cp --recursive "$SMDEBUG_S3_BINARY" s3_pip_binary
```
Comment (Vikas-kum) [inline config/install_smdebug.sh]: check_smdebug_install - How is this knowing what version to check ? 

$force_check_smdebug_version - this variable controls , see config/check_smdebug_install.sh . We need to make sure force_check_smdebug_version this is set to right version from pip binary.

File: config/buildspec_build_wheel.yml (hunk 1 above)
Comment (anirudhacharya) [reply config/buildspec_build_wheel.yml]: these are already declared as env variables in the codebuild project. I can also include it as env variables in the buildspec before the phases.

File: config/buildspec_zero_code_change.yml (hunk 3 above)
Comment (anirudhacharya) [reply config/buildspec_zero_code_change.yml]: only line 33 was removed. because the installs happens on line 32 from the install_smdebug script.

File: config/buildspec_zero_code_change.yml (hunk 2 above)
Comment (anirudhacharya) [reply config/buildspec_zero_code_change.yml]: pip wheels are built from master branch. giving an option here to choose a different branch might cause inconsistencies.

File: config/buildspec_zero_code_change.yml (hunk 4 above)
Comment (anirudhacharya) [reply config/buildspec_zero_code_change.yml]: the build project for this buildspec is only run as part of the nightly pipeline. So this will anyway get overridden by pipeline settings and all the logs will get uploaded to the s3 bucket specified by the pipeline artifacts. So i thought I will remove it. Can put it back, but it should not make any difference.

File: config/install_smdebug.sh (hunk 5 above)
Comment (anirudhacharya) [reply config/install_smdebug.sh]: > We need to make sure force_check_smdebug_version this is set to right version from pip binary.

i will do this.

File: config/buildspec_zero_code_change.yml (hunk 4 above)
Comment (Vikas-kum) [reply config/buildspec_zero_code_change.yml]: One use case if we want to run specific tests like TF tests for certain PRs which changes and affects TF ZCC . 
We should have a way to run these test in isolation outside of pipeline, if it manually triggers.

File: config/buildspec_build_wheel.yml [hunk 6]
```diff
@@ -0,0 +1,42 @@
...
+      - cd $CODEBUILD_SRC_DIR  && echo $CORE_COMMIT > smdebug/CORE_COMMIT
+      - cd $CODEBUILD_SRC_DIR && git checkout $CORE_COMMIT && python setup.py bdist_wheel --universal
+      - cd $CODEBUILD_SRC_DIR_RULES && if [ -z "$RULES_COMMIT" ]; then export RULES_COMMIT=$(git log -1 --pretty=%h); fi
+      - cd $CODEBUILD_SRC_DIR_RULES  && echo $RULES_COMMIT > smdebug_rules/RULES_COMMIT
+      - cd $CODEBUILD_SRC_DIR_RULES && python setup.py bdist_wheel --universal
```
Comment (Vikas-kum) [inline config/buildspec_build_wheel.yml]: No git checkout for RULES_COMMIT ?

File: config/install_smdebug.sh [hunk 7]
```diff
@@ -2,17 +2,36 @@
...
+git clone "$RULES_REPO"
+
+export CODEBUILD_SRC_DIR=`pwd`/sagemaker-debugger
+export CODEBUILD_SRC_DIR_RULES=`pwd`/sagemaker-debugger-rules
+
+echo "sagemaker-debugger repository cloned in the path $CODEBUILD_SRC_DIR"
+echo "sagemaker-debugger-rules repository cloned in the path $CODEBUILD_SRC_DIR_RULES"
 
 cd $CODEBUILD_SRC_DIR
-# you can provide bip binary as s3 path in the build environment
```
Comment (Vikas-kum) [inline config/install_smdebug.sh]: We want both ways - 
1) Running in pipeline, which would work on binary
2) Just run build on individual branch and commit, which would require existing way.

File: config/buildspec_zero_code_change.yml (hunk 4 above)
Comment (anirudhacharya) [reply config/buildspec_zero_code_change.yml]: ok. will change accordingly.
